ludwig/api.py

```diff
@@ -983,31 +994,90 @@ def generate(
+
+            decoded_outputs = tokenizer.tokenizer.batch_decode(outputs, skip_special_tokens=True)
+            logger.info(f"Finished generating in: {(time.time() - start_time):.2f}s.")
+
+            return decoded_outputs[0] if len(decoded_outputs) == 1 else decoded_outputs
+
+    def _generate_streaming_outputs(
+        self,
+        input_strings: Union[str, List[str]],
+        input_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        streamer: Union[TextStreamer, None],
+    ):
+        """Generate streaming outputs for the given input.
+
+        Args:
+            input_strings (Union[str, List[str]]): Input text or list of texts to generate from.
+            input_ids (torch.Tensor): Tensor containing input IDs.
+            attention_mask (torch.Tensor): Tensor containing attention masks.
+            streamer (Union[TextStreamer, None]): Text streamer instance for streaming output.
+
+        Returns:
+            torch.Tensor: Concatenated tensor of generated outputs.
+        """
+        outputs = []
+        input_strings = input_strings if isinstance(input_strings, list) else [input_strings]
+        for i in range(len(input_ids)):
             with torch.no_grad():
-                outputs = self.model.model.generate(
-                    input_ids=input_ids,
-                    attention_mask=attention_mask,
-                    # NOTE: self.model.model.generation_config is not used here because it is the default
-                    # generation config that the CausalLM was initialized with, rather than the one set within the
-                    # context manager.
+                logger.info(f"Input: {input_strings[i]}\n")
+                # NOTE: self.model.model.generation_config is not used here because it is the default
+                # generation config that the CausalLM was initialized with, rather than the one set within the
+                # context manager.
+                generated_output = self.model.model.generate(
+                    input_ids=input_ids[i].unsqueeze(0),
+                    attention_mask=attention_mask[i].unsqueeze(0),
                     generation_config=self.model.generation,
+                    streamer=streamer,
                 )
-                decoded_outputs = tokenizer.tokenizer.batch_decode(outputs, skip_special_tokens=True)
-                logger.info(f"Finished generating in: {(time.time() - start_time):.2f}s.")
-                if len(decoded_outputs) == 1:
-                    return decoded_outputs[0]
-                return decoded_outputs
+                logger.info("----------------------")
+                outputs.append(generated_output)
+        return torch.cat(outputs, dim=0)
+
+    def _generate_non_streaming_outputs(
+        self,
+        _input_strings: Union[str, List[str]],
+        input_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        streamer: Union[TextStreamer, None],
```

Comment:
@arnavgarg1 So here, we would not include the `streamer`, because we know that it is `None` -- and thus pass `None` to `self.model.generate()`?  Is this reasonable?

Reply:
It is reasonable since it defaults to None by default in the generate() call in the model API, but I cleaned this up so that we only pass in the argument into generate when the streamer is actually present. Let me know if that helps!